src/samplerate.rs

```diff
@@ -95,29 +95,58 @@ impl Samplerate {
         }
     }
 
-    /// Perform a samplerate conversion of the given input data.
-    /// If the number of channels used was not `1` (Mono), the samples are expected to be stored
-    /// interleaved.
-    pub fn process(&self, input: &[f32]) -> Result<Vec<f32>, Error> {
+    fn _process(&self, input: &[f32], output_len: usize, end_of_input: bool) -> Result<Vec<f32>, Error> {
         let channels = self.channels()?;
-        let output_len = (self.ratio() * input.len() as f64) as usize;
         let mut output = vec![0f32;output_len];
         let mut src = SRC_DATA {
             data_in: input.as_ptr(),
             data_out: output.as_mut_ptr(),
             input_frames: (input.len() as i32 / channels as i32).into(),
             output_frames: (output_len as i32 / channels as i32).into(),
             src_ratio: self.ratio(),
-            end_of_input: 0,
+            end_of_input: if end_of_input { 1 } else { 0 },
             input_frames_used: 0,
             output_frames_gen: 0,
+            ..Default::default()
         };
         let error_int = unsafe { src_process(self.ptr, &mut src as *mut SRC_DATA) };
         match ErrorCode::from_int(error_int) {
-            ErrorCode::NoError => Ok(output),
+            ErrorCode::NoError => Ok(output[..src.output_frames_gen as usize*channels].into()),
             _ => Err(Error::from_int(error_int)),
         }
     }
+
+    /// Perform a samplerate conversion on a block of data (use `process_last` if it is the last one)
+    /// If the number of channels used was not `1` (Mono), the samples are expected to be stored
+    /// interleaved.
+    pub fn process(&self, input: &[f32]) -> Result<Vec<f32>, Error> {
+        self._process(input, (self.ratio() * input.len() as f64) as usize, false)
+    }
+    
+    /// Perform a samplerate conversion on last block of given input data.
+    /// If the number of channels used was not `1` (Mono), the samples are expected to be stored
+    /// interleaved.
+    pub fn process_last(&self, input: &[f32]) -> Result<Vec<f32>, Error> {
+        let output_len = (self.ratio() * input.len() as f64) as usize;
+        match self._process(input, output_len, true) {
+            Ok(mut output) => {
+                loop {
+                    match self._process(&[0f32; 0], output_len, true) {
```

Comment:
What is the purpose of this? I don't understand it.
You're processing once (above) the input with the given length. Then You're taking the output and convert a set of zeroes until the stream is consumed. And you write these chunks (will they not always be zeroes?) into the end of the output.

Do I understand this correctly? Are you doing this to fix a problem where the output length is unexpectedly small?

Reply:
Is this a bug you're fixing? If yes, then shouldn't this be called always when calling `process`?

Reply:
I don't convert a stream of zeroes, I call `src_process` with no input (`[0f32; 0]` has length 0) and `end_of input` set to 1, to get the last few samples. The remaining samples are there because `src_process` won't return more samples then `input*ratio` but it can return less, at which point the remaining samples are stored in it's internal buffer.

The user has to call it when he knows that the block he is processing is the last one to get all samples (It's not required of course to call it if he doesn't care about a few lost samples)

See here:
>You are supposed to keep calling `src_proccess` with zero input samples and the `end_of_input`  field set to `1` until you get not more output samples.
The examples do this, for example `examples/timewarp-file.c`.

_Originally posted by @erikd in https://github.com/erikd/libsamplerate/issues/55#issuecomment-479433300_